### DISCUSSION TITLE
Conditional Assignment Operations in engine.Eval

### DISCUSSION THREAD

**Discussion by cchannon** [microsoft/Power-Fx#815] [Q&A] [answered]
I'm working on a PowerFX implementation for handling relatively straightforward business logic units. I expect users to need to perform assignment (Set) operations primarily and to perform them conditionally.

I've reviewed the samples closely and I've played around with the RecalcEngine a bit. So far, I am a bit surprised by how challenging this objective seems to be. The RecalcEngine can't handle Set inside a conditional expression (it says Set is an invalid Function) and the samples seem to bear this out, with the REPL sample openly assuming that Set can only be evaluated if it is the opening of a REPL command (The regex match begins with @"^\s*Set).

So, from what I see so far, what I need to do is Regex match every Set operation and substitute them with some Output value like Set1, Set2, etc. Then eval the expression and match my Output back to the Set operation I swapped out, then run that Set operation. Something kinda like the below:

![image](https://user-images.githubusercontent.com/12833126/200325255-d3c227dd-4797-4f24-9038-696bad1684c9.png)

But that doesn't seem right to me. In the Canvas App editor, there is no problem nesting an If inside a Set inside an If, so I feel like I must be missing something here, I just can't seem to find it.

**Comment by anderson-joyle** [comment, marked as answer]
Power Fx requires explicitly enabling because hosts generally want control of what the expressions can do. A lot of hosts want to explicitly constrain to be trivial "side-effect" free expressions.
Please take a look at the following test cases on how to enable Set function and how Power Fx engine deals with mutation functions.
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Interpreter.Tests/SetFunctionTests.cs

**Comment by cchannon** [comment]
OK, Awesome! Yes, this is getting me MUCH further along - thank you!

So let me run this by you and see if I'm getting it right here. If we look at the [REPL example](https://github.com/microsoft/power-fx-host-samples/blob/main/Samples/ConsoleREPL/ConsoleREPL.cs) then there is a simple code block there for recognizing Set() and updating a var without using Eval because the example does not set the EnableSetFunction flag...

```
  if ((match = Regex.Match(expr, @"^\s*Set\(\s*(?<ident>\w+)\s*,\s*(?<expr>.*)\)", RegexOptions.Singleline)).Success)
  {
      var r = engine.Eval(match.Groups["expr"].Value);
      Console.WriteLine(match.Groups["ident"].Value + ": " + PrintResult(r));
      engine.UpdateVariable(match.Groups["ident"].Value, r);
  }
```

This basic approach is just so people can see it is possible, but a more mature approach has to set the EnableSetFunction flag, then inspect each expr as it comes through to search for Set functions, then check to make sure the ident they are declaring already exists (can't Set an ident that hasn't already been declared). So, then, something more functional would look kinda like this:

```
...
config.EnableSetFunction();
engine = new RecalcEngine(config);
...
MatchCollection allMatches = Regex.Matches(expr, @"\s*Set\(\s*?(?<ident>\w+?)\s*?,\s*?(?<expr>.*?)\)", RegexOptions.Singleline);
if(allMatches.Count>0)
{
    foreach(Match thisMatch in allMatches)
    {
        try
        {
            var val = engine.GetValue(thisMatch.Groups["ident"].Value);
        }
        catch
        {
            var r = engine.Eval(thisMatch.Groups["expr"].Value);
            engine.UpdateVariable(thisMatch.Groups["ident"].Value, FormulaValue.NewBlank(r.Type));
        }
    }
    engine.Eval(expr, null, _opts);
} 
```

Feels like I'm getting there now...

**Reply by cchannon** [reply]
Or, I guess, I could just skip the foreach and just jump directly to the Eval in its own trycatch, then inspect the exception to find the missing key... but then I wouldn't have the correct Type for that key value, so, yeah, I guess I stick with this loop...🤔

**Reply by MikeStall** [reply]
We should definitely remove that Set() regex from the repl.  That was added from before we had a real Set() function available.

**Reply by cchannon** [reply]
I'll just PR the edit from above if it works for you

**Reply by cchannon** [reply]
I'm still not totally happy with the Regex because it doesn't properly handle functions within a Set but TBH all this Regex is making me dizzy

**Reply by cchannon** [reply]
If I might suggest, a dev-friendly feature might be to enhance .Check (or introduce a second kind of Check) that inspects the expr and returns _runtime-convenient_ metadata (as opposed to the _authoring-time_ information I have seen in there so far) such as any identities declared, their implied types, the inferred operation type, etc. But... Maybe that's already in there and I just haven't seen it yet... Still getting started!

**Reply by anderson-joyle** [reply]
Take a look at these test cases:
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Core.Tests/BindingEngineTests.cs

They call **Engine.Check(expression)** (not to confuse with RecalcEngine). It verifies any syntax error and tries to resolve the identifiers.
The result is a CheckResult object containing the parse tree. From there you can traverse the nodes and get the details you are looking for.

**Reply by cchannon** [reply]
The .parse method alone! 🤯

OK, I will dig into this more closely, then I will send y'all a PR with a nice, clean rewrite of the ConsoleREPL sometime in the next couple days. 

Is there a documented error list somewhere that can help me _parse_ these results? 😁

**Reply by cchannon** [reply]
Ooooooohhhh the error "number" is the error range of index positions - I get it now! You're giving _both_ the authoring validation and the runtime validation. very nice!
